cvat/apps/engine/static/engine/js/shapeCreator.js

```diff
@@ -954,10 +957,105 @@ class ShapeCreatorView {
                     this._aim.x.attr('stroke-width', STROKE_WIDTH / this._scale);
                     this._aim.y.attr('stroke-width', STROKE_WIDTH / this._scale);
                 }
-                if (['box', 'polygon', 'polyline'].includes(this._type)) {
+                if (['box', 'polygon', 'polyline', 'rays'].includes(this._type)) {
                     this._drawInstance.attr('stroke-width', STROKE_WIDTH / this._scale);
                 }
             }
         }
     }
 }
+
+class RaysDrawInstance {
+    constructor(creatorView) {
+        this._creatorView = creatorView;
+        this._finishedRays = [];
+        this._currentRay = null;
+        this._isRayInProgress = false;
+        this._beginRay();
+    }
+
+    draw(type) {
+        if (type === 'done') {
+            this._finish();
+        } else if (type === 'cancel') {
+            this._delegate(el => el.draw('cancel'));
+        } else {
+            throw new Error(`Event ${type} is not supported`);
+        }
+    }
+
+    remove(){
+        this._delegate(el => el.remove());
+    }
+
+    attr(key, value) {
+        this._delegate(el => el.attr(key, value));
+    }
+
+    _delegate(callback) {
+        if (this._currentRay) {
+            callback(this._currentRay);
+        }
+        this._finishedRays.forEach(r => callback(r));
+    }
+
+    _beginRay() {
+        this._currentRay = this._creatorView._frameContent.polyline().draw({snapToGrid: 0.1})
+            .addClass('shapeCreation').attr({
+                'stroke-width': STROKE_WIDTH / this._creatorView._scale,
+            });
+        this._currentRay.attr({
+            z_order: Number.MAX_SAFE_INTEGER,
+        });
+
+        this._currentRay.on('drawstart', (e) => {
+            this._isRayInProgress = true;
+        });
+
+        this._currentRay.on('drawpoint', (e) => {
+            this._currentRay.draw("done");
+            this._isRayInProgress = false;
+            this._finishRay();
+        });
+
+        this._currentRay.on('drawstart', this._creatorView._rescaleDrawPoints.bind(this._creatorView));
+        this._currentRay.on('drawpoint', this._creatorView._rescaleDrawPoints.bind(this._creatorView));
+
+        this._currentRay.on('drawstop', () => {
+            this._creatorView._frameContent.off('mousedown.shapeCreator');
+            this._creatorView._frameContent.off('mousemove.shapeCreator');
+            $('body').off('keydown.shapeCreator');
+        });
+    }
+
+    _finishRay() {
+        this._finishedRays.push(this._currentRay);
+        if (this._creatorView._polyShapeSize && this._finishedRays.length >= this._creatorView._polyShapeSize) {
+            this._finish();
+        } {
+            this._beginRay();
+        }
+    }
+
+    _finish() {
+        const { arePointsEqual, findVanishingPoint } = window.graphicPrimitives;
+        let actualPoints = this._finishedRays.map(e => e.attr('points')).join(" ");
```

Comment:
why do you need string conversion here?
why is it not JSON if you need string representation?

Reply:
CVAT stores figures coordinates in string representation and all CVAT functions assume that figures are stored that way, in order to use those functions I have to store coordinates the same way, otherwise I have to re-implement a lot of existing functionality.

Here I join coordinates of multiple lines into one polyline (`"10372.84,10045.72 10277.81,9435.34" "9894.04,10027.44 9996.38,9486.51" "9499.30,9819.11 9788.04,9471.89"` => `"10372.84,10045.72 10277.81,9435.34 9894.04,10027.44 9996.38,9486.51 9499.30,9819.11 9788.04,9471.89"`).